### PR TITLE
[hail/ptypes] non-staged copyFromType

### PR DESCRIPTION
**Pull request by akotlar**
Stacked on https://github.com/hail-is/hail/pull/7639

Tim, the problem I mentioned was solved. The issue was that setElementMissing was unfortunately calling the staged implementation, because of Scala implicit auto-coercision from Long to Code[Long]. :(

### CODE REVIEW

File: hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala [hunk 1]
```diff
@@ -71,6 +71,9 @@ trait PArrayBackedContainer extends PContainer {
   def setElementMissing(region: Region, aoff: Long, i: Int) =
     arrayRep.setElementMissing(region, aoff, i)
 
+  def setElementMissing(aoff: Long, i: Int) =
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PArrayBackedContainer.scala]: This makes the staged and unstaged setElementMissing have constructors with same arity (namely sans region). I can hold off on this change if you prefer, for now.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala [hunk 2]
```diff
@@ -367,4 +386,140 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
...
+    val numberOfElements = sourceType.loadLength(srcAddress)
+    val dstAddress = this.allocate(region, numberOfElements)
+    this.initialize(region, dstAddress, numberOfElements)
+
+    var currentElementAddress = this.firstElementOffset(dstAddress, numberOfElements)
+    var currentIdx = 0
+
+    while(currentIdx < numberOfElements) {
+      if(!sourceElementType.required && sourceType.isElementMissing(region, srcAddress, currentIdx)) {
+        this.setElementMissing(dstAddress, currentIdx)
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: unfortunately, calling this without `region` as the first argument, caused the staged version to be dispatched, due to implicit conversion!

File: hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala [hunk 3]
```diff
@@ -191,7 +215,10 @@ abstract class PBaseStruct extends PType {
   def isFieldDefined(region: Code[Region], offset: Code[Long], fieldIdx: Int): Code[Boolean] =
     isFieldDefined(offset, fieldIdx)
 
-  def setFieldMissing(region: Region, offset: Long, fieldIdx: Int) {
+  def setFieldMissing(region: Region, offset: Long, fieldIdx: Int): Unit =
+    setFieldMissing(offset, fieldIdx)
+
+  def setFieldMissing(offset: Long, fieldIdx: Int) {
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/expr/types/physical/PBaseStruct.scala]: same issue and solution as setElementMissing.

File: hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala (hunk 2 above)
Comment (tpoterba) [reply hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala]: ack!! those implicit conversions...

**Comment by akotlar** [comment]
Ready to review

**Pull request merged**